File: Leetcode/690. Employee Importance/solution.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""


class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        employees = {employee.id: employee for employee in employees}

        def dfs(root_employee):
            if not root_employee:
                return 0
            return root_employee.importance + sum([
                dfs(employees.get(subordinate)) for subordinate in root_employee.subordinates])

        return dfs(employees.get(id))

File: Leetcode/690. Employee Importance/solution.cpp
#include <unordered_map>

/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
public:
    std::unordered_map<int, Employee*> employeesMap;
    
    int getImportance(vector<Employee*> employees, int id) {
        for(auto& emp : employees){
            employeesMap[emp->id] = emp;
        }

        return dfs(employeesMap[id]);
    }
    
    int dfs(Employee* rootEmployee) {
        if (rootEmployee == NULL) return 0;
        
        vector<int> subsImportance;
        
        for (auto& subId : rootEmployee->subordinates) {
            subsImportance.push_back(dfs(employeesMap[subId]));
        }

        return rootEmployee->importance + accumulate(subsImportance.begin(), subsImportance.end(), 0);
    }
};